### ISSUE TITLE
[bug] Issue with app.handle().exit(0) Not Working When Running as Windows Service

### ISSUE THREAD

**Issue by arihav** [tauri-apps/tauri#11573]
### Describe the bug

I am currently running my Tauri app as a Windows service (without a window), and I am encountering an issue when trying to close the app using the app.handle().exit(0) command.

### Reproduction

https://github.com/arihav/tauri-app-example

### Expected behavior

The app should exit successfully without any issues.

### Full `tauri info` output

```text
[✔] Environment
    - OS: Windows 10.0.22631 x86_64 (X64)
    ✔ WebView2: 130.0.2849.56
    ✔ MSVC: Visual Studio Build Tools 2022
    ✔ rustc: 1.79.0 (129f3b996 2024-06-10)
    ✔ cargo: 1.79.0 (ffa9cf99a 2024-06-03)
    ✔ rustup: 1.27.1 (54dd3d00f 2024-04-24)
    ✔ Rust toolchain: stable-x86_64-pc-windows-msvc (environment override by RUSTUP_TOOLCHAIN)
    - node: 21.1.0
    - yarn: 1.22.22
    - npm: 10.2.0

[-] Packages
    - tauri 🦀: git+https://github.com/tauri-apps/tauri?rev=ae024b829972609ac565e54bf4b03d4e4b5e2524#ae024b829972609ac565e54bf4b03d4e4b5e2524 (2.0.6)
    - tauri-build 🦀: No version detected
    - wry 🦀: 0.46.3
    - tao 🦀: 0.30.3
    - tauri-cli 🦀: 2.0.4
    - @tauri-apps/api : 1.6.0 (outdated, latest: 2.0.3)
    - @tauri-apps/cli : 2.0.4

[-] Plugins
    - tauri-plugin-log 🦀: 2.0.1
    - @tauri-apps/plugin-log : 2.0.0
    - tauri-plugin-autostart 🦀: 2.0.1
    - @tauri-apps/plugin-autostart : 2.0.0
    - tauri-plugin-localhost 🦀: 2.0.1
    - @tauri-apps/plugin-localhost : not installed!

[-] App
    - build-type: bundle
    - CSP: unset
    - frontendDist: ../out
    - devUrl: http://localhost:3002/
    - framework: React (Next.js)
    - bundler: Webpack
```

### Additional context

The exit process should trigger two events: RunEvent::ExitRequested and RunEvent::Exit. Currently, RunEvent::ExitRequested is called, but RunEvent::Exit is not. I think the issue might be occurring within the event loop of TAO.

**Comment by Legend-Master** [comment]
Your example code shows that you're blocking the setup hook when calling `app.exit`, does it still happen if you use `spawn` instead of `spawn_blocking` here?

https://github.com/arihav/tauri-app-example/blob/0dbd487aa4323ceacf299ea9fe8c8bc2d7adbe74/src-tauri/src/lib.rs#L18-L21

**Comment by arihav** [comment]
Thank you for the suggestion! I tried using spawn instead of spawn_blocking, but the issue persists.

**Comment by Legend-Master** [comment]
Just realized that you're trying to run the tauri app as windows service, this will not work, you can't run `tauri::Builder::default().run()` in that environment

I'm not sure what you want to achieve here, since a service is not suppose to have GUIs, if you want your app to run as admin or you want the service to spawn your app in some conditions, see https://learn.microsoft.com/en-us/windows/win32/services/interactive-services

**Comment by ari-hav** [comment]
In the example, I have a window with a GUI, but in the actual use case, I don’t open a visible window. Instead, I use the localhost plugin to open a port for the UI, allowing another app to integrate my app as an iframe.

**Comment by arihav** [comment]
In the example, I have a window with a GUI, but in the actual use case, I don’t open a visible window. Instead, I use the localhost plugin to open a port for the UI, allowing another app to integrate my app as an iframe.

**Comment by Legend-Master** [comment]
In that case just don't use localhost plugin, setup the localhost server with things like reqwest instead

Also, just curios, did you get `tauri::Builder::default().run()` running when it's running as windows service? It just panicked when I tried it

**Comment by arihav** [comment]
Thanks for the suggestion! Could you explain why it might be better to use another server setup instead of the `localhost` plugin? I’d love to understand the benefits or differences.

Also, I had an issue in the initial example, but I just pushed a fix. The app now runs on port `1420` when running as a Windows service. Thanks for your help!

https://github.com/arihav/tauri-app-example

For reinstalling the update, you can kill the app in Task Manager.

**Comment by Legend-Master** [comment]
I'm still kinda confused what you're trying to do, the localhost plugin runs a server to serve the files bundled inside the tauri exe file, I'm thinking if you're only using the localhost plugin and tauri to serve static files, why not just use reqwest to serve local files instead?

I can launch the your example now, trying to trace a bit more, I suspect the cause is this line https://github.com/tauri-apps/tao/blob/aff33fbb2dab66cca2481fb75a3a6d612689269c/src/platform_impl/windows/event_loop.rs#L268

**Comment by arihav** [comment]
Thanks for your input! I checked on your suggestion regarding using reqwest to serve local files instead of the localhost plugin.

Regarding the second point, I also traced the error to the line you mentioned in the event loop. I noticed that the `move_state_to` function is never called with `HandlingRedrawEvents`, unlike in a regular run.

**Comment by Legend-Master** [comment]
Yeah, from debugging, it seems like normally the state will be changed to `Idle` on `WM_PAINT` event, but in the service context, it's never emitted, which made sense

I'm not familiar with how this part works to be honest, it's from winit years ago and they seem to also moved away from it at the end of last year

This is not the intended use case to be honest, if you really want to do this, try using `app.cleanup_before_exit()` and then `std::process::exit(exit_code)` instead to skip all the event loop stuffs and just shutdown

**Comment by ari-hav** [comment]
Thx, that's what I'll do.

**Comment by arihav** [comment]
Thx, that's what I'll do.

**Comment by ahqsoftwares** [comment]
Unless the op targets windows xp / 7 / 8.1, I don't think interactive services work anyways
